gc collect from a gc_thread

yancanmao opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Ray proactively triggers gc.collect() on idle workers to release Python objects that may still hold Plasma shared memory (shm) references.
In the current implementation in (_raylet.pyx gc_collect()), Ray calls gc.collect() from Cython under a with gil block periodically.
If the Python object graph is complex (e.g., cyclic references with finalizers), gc.collect() may take a long time. During this period, since the GIL is held for the entire collection, user code is completely frozen if gc.collect() time is longer than the periodic interval (e.g., 10s).

We propose decoupling GC execution from the RPC call:

gc_collect in Cython should not directly run gc.collect().
Instead, it should "signal an event" with minimum execution time (e.g., using a threading.Event or similar).
A dedicated Python GC thread consumes this event and executes gc.collect() asynchronously, with a configurable GC interval.

## Related issue number

Closes #55837

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Commit:
gc collect from gc_thread

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

Commit:
code clean

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+_gc_lock = threading.Lock()
+
+
+class GCManagerThread(threading.Thread):
```
```suggestion
class PythonGCThread(threading.Thread):
```
update the `name` below to match. this will help make it clear to someone who dumps all of the threads in the process what it does

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+                self.gc_in_progress = True
+                try:
+                    start = time.perf_counter()
+                    num_freed = gc.collect()
```
Good points, let's go ahead with the proposed change

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+        self.join()
+
+
+def init_gc_collect_manager(min_interval: int = 5) -> None:
+    global _gc_thread
+    with _gc_lock:
+        if _gc_thread is not None and _gc_thread.is_alive():
+            logger.warning("GC thread already initialized")
+            return
+
+        _gc_thread = GCManagerThread(min_interval)
+        _gc_thread.start()
+        logger.debug(f"GC thread initialized with min interval {min_interval}s")
+
+
+def stop_gc_collect_manager_if_needed() -> None:
+    global _gc_thread
+    with _gc_lock:
+        if _gc_thread is not None:
+            _gc_thread.stop()
+            _gc_thread = None
+            logger.debug("GC thread stopped")
+        else:
+            logger.debug("No GC thread to stop")
```
I think these can be simplified to be started in the `CoreWorker` constructor in `_raylet.pyx`. no need for concurrency control or lazy initialization

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+            logger.debug("No GC thread to stop")
+
+
+def trigger_gc() -> None:
+    _gc_event.set()
```
we can make this a method on the `GCManagerThread` and have the core worker call it directly (if it keeps the `GCManagerThread` as an attribute)

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+                                num_freed, self._last_gc_time - start
+                            )
+                        )
+                except Exception as e:
+                    logger.error(f"Error during GC: {e}")
```
we should update the last_gc_time in this case to avoid hotlooping

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,86 @@
+                )
+                continue
+
+            if not self.gc_in_progress:
+                self.gc_in_progress = True
```
what's the purpose of `gc_in_progress`?

edoakes commented on python/ray/_raylet.pyx:
```diff
@@ -3047,6 +3054,11 @@ cdef class CoreWorker:
         self._task_id_to_future = {}
         self.event_loop_executor = None
 
+        if RayConfig.instance().start_python_gc_manager_thread():
+            init_gc_collect_manager(ray_constants.RAY_GC_MIN_COLLECT_INTERVAL)
```
let's make the thread an attribute of the core worker (see my other comment)

we also need to make sure to stop & join this thread when the core worker is shut down (for example if the user calls `ray.shutdown` explicitly, the thread should be stopped & joined)

edoakes commented on python/ray/tests/test_global_gc.py:
```diff
@@ -216,5 +217,82 @@ def f(self):
         gc.enable()
 
 
+@pytest.mark.timeout(30)
+@pytest.mark.parametrize("support_fork", [True, False])
+def test_long_local_gc(shutdown_only, support_fork):
```
let's add a unit test for the thread implementation as well. you can dependency inject a fake `gc.collect` implementation and check that the event, joining, etc. logic works correctly

yancanmao committed:
Update python/ray/_private/gc_collect_manager.py

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Mao Yancan <maoyancan@u.nus.edu>

Commit:
Revision based on comments

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

yancanmao commented:
Thanks for your insightful comments! I have revised the code as follows:
1. Renamed `GCThreadManager` to `PythonGCThread`.
2. Made `trigger_gc()` a method inside `PythonGCThread`.
3. Removed locks/variables for concurrency control.
4. Added UT for gc thread tests.
5. Added `gc_thread.stop()` inside `shutdown_driver()`

Commit:
lint reformat

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

Commit:
fix ut

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,64 @@
+
+
+class PythonGCThread(threading.Thread):
+    """A background thread that triggers Python garbage collection.
+    This thread waits for GC events from CoreWorker and triggers `gc.collect()` when
```
```suggestion
    """A background thread that triggers Python garbage collection.

    This thread waits for GC events from CoreWorker and triggers `gc.collect()` when
```
nit

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,64 @@
+    `min_interval` seconds."""
+
+    def __init__(
+        self, min_interval: int = 5, gc_collect_func: Optional[Callable] = None
```
```suggestion
        self, *, min_interval_s: int = 5, gc_collect_func: Optional[Callable] = None
```
nits:

- force args to be passed as kwargs
- use unit in the arg name for clarity

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,64 @@
+    ):
+        logger.debug("Starting Python GC thread")
+        super().__init__(name="PythonGCThread", daemon=True)
+        self._running = True
```
```suggestion
        self._should_exit = False
```
nit: a little clearer IMO

edoakes commented on python/ray/_private/gc_collect_manager.py:
```diff
@@ -0,0 +1,64 @@
+            if not self._running:
+                break
+
+            time_since_last_gc = time.perf_counter() - self._last_gc_time
```
no need to use `perf_counter` here -- it's more expensive since it's high resolution

edoakes commented on python/ray/tests/test_global_gc.py:
```diff
@@ -216,5 +219,140 @@ def f(self):
+        assert gc_thread.is_alive()
+
+        gc_thread.trigger_gc()
+        time.sleep(0.1)
```
we shouldn't depend on timing conditions in tests -- please inject a fake clock instead of using real `time.sleep` and make the tests deterministic

edoakes commented on python/ray/tests/test_global_gc.py:
```diff
@@ -216,5 +219,140 @@ def f(self):
 
+@pytest.mark.timeout(30)
+@pytest.mark.parametrize("support_fork", [True, False])
+def test_long_local_gc(shutdown_only, support_fork):
```
This test is too expensive/slow to run in CI and is highly likely to be flaky. Instead, let's add a basic test to check that the garbage is only collected once per interval, and set the interval to be something modest like 5s.

edoakes commented:
Mostly nits at this point, thanks for addressing the comments!

yancanmao committed:
Update python/ray/_private/gc_collect_manager.py

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Mao Yancan <maoyancan@u.nus.edu>

yancanmao committed:
Update python/ray/_private/gc_collect_manager.py

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Mao Yancan <maoyancan@u.nus.edu>

yancanmao committed:
Update python/ray/_private/gc_collect_manager.py

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Mao Yancan <maoyancan@u.nus.edu>

Commit:
nits update

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

Commit:
lint update

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

yancanmao committed:
Merge branch 'master' into new_gc_collect_thread

Signed-off-by: Mao Yancan <maoyancan@u.nus.edu>

Commit:
gc_thread use monotonic instead of perf_counter

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

Commit:
Merge branch 'new_gc_collect_thread' of https://github.com/yancanmao/ray into new_gc_collect_thread

Commit:
lint reformat

Signed-off-by: Mao Yancan <yancan.mao@bytedance.com>

yancanmao commented:
> Mostly nits at this point, thanks for addressing the comments!

Thanks for the comments!  I have updated the code accordingly.

edoakes commented:
I've triggered CI. Please tag me when tests are passing for merge!

edoakes committed:
Merge branch 'master' into new_gc_collect_thread
